### PR TITLE
add range search for hgraph (only interface)

### PR DESCRIPTION
**Pull request by LHT129**
issue #40 
- introduce InnerSearchParam for hgraph to avoid too much passing params
- introduce enum InnerSearchMode for hgraph, for knn search or range search

### CODE REVIEW

**Review by inabao — APPROVED** [review]
LGTM

File: src/index/hgraph_index.h [hunk 1]
```diff
@@ -83,33 +83,32 @@ class HGraphIndex : public Index {
...
     }
 
-    // TODO(LHT): implement
     tl::expected<DatasetPtr, Error>
     RangeSearch(const DatasetPtr& query,
                 float radius,
                 const std::string& parameters,
                 int64_t limited_size = -1) const override {
-        return Dataset::Make();
+        SAFE_CALL(return this->range_search(query, radius, parameters, nullptr, limited_size););
```
Comment (jiaweizone) [inline src/index/hgraph_index.h]: `limited_size););` has compile issue ?

File: src/index/hgraph_index.h [hunk 2]
```diff
@@ -83,33 +83,32 @@ class HGraphIndex : public Index {
...
-    // TODO(LHT): implement
     tl::expected<DatasetPtr, Error>
     RangeSearch(const DatasetPtr& query,
                 float radius,
                 const std::string& parameters,
                 BitsetPtr invalid,
                 int64_t limited_size = -1) const override {
-        return Dataset::Make();
+        BitsetOrCallbackFilter filter(invalid);
+        SAFE_CALL(return this->range_search(query, radius, parameters, &filter, limited_size););
```
Comment (jiaweizone) [inline src/index/hgraph_index.h]: ditto

File: src/index/hgraph_index.h [hunk 3]
```diff
@@ -83,33 +83,32 @@ class HGraphIndex : public Index {
...
-    // TODO(LHT): implement
     tl::expected<DatasetPtr, Error>
     RangeSearch(const DatasetPtr& query,
                 float radius,
                 const std::string& parameters,
                 const std::function<bool(int64_t)>& filter,
                 int64_t limited_size) const override {
-        return Dataset::Make();
+        BitsetOrCallbackFilter callback(filter);
+        SAFE_CALL(return this->range_search(query, radius, parameters, &callback, limited_size););
```
Comment (jiaweizone) [inline src/index/hgraph_index.h]: ditto

File: src/index/hgraph_index.h (hunk 1 above)
Comment (LHT129) [reply src/index/hgraph_index.h]: no problem, have been compiled successful

File: src/index/hgraph_index.h [hunk 4]
```diff
@@ -83,33 +83,32 @@ class HGraphIndex : public Index {
...
     }
 
-    // TODO(LHT): implement
     tl::expected<DatasetPtr, Error>
     RangeSearch(const DatasetPtr& query,
                 float radius,
                 const std::string& parameters,
                 int64_t limited_size = -1) const override {
-        return Dataset::Make();
+        SAFE_CALL(return this->range_search(query, radius, parameters, nullptr, limited_size);)
```
Comment (jiaweizone) [inline src/index/hgraph_index.h]: SAFE_CALL(return this->range_search(query, radius, parameters, nullptr, limited_size));

Comment (LHT129) [reply src/index/hgraph_index.h]: done

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**